sklearn/tree/_criterion.pyx

```diff
@@ -599,6 +600,95 @@ cdef class Gini(ClassificationCriterion):
+    If k represents a node before split, k1 and k2 represent the children nodes after split,
+    left_0 represents observations of class 0 in node k1, right_0 represents observations of class 0 in node k2,
+    left_1 represents observations of class 1 in node k1, right_1 represents observations of class 1 in node k2,
+    parent_0 represents observations of class 0 in node k, parent_1 represents observations of class 1 in node k,
+    then let
+
+        count_k1 = (\sqrt{\frac{N_{left_0}}{N_{parent_0}}}-\sqrt{\frac{N_{left_1}}{N_{parent_1}}})^2
+
+    be the partial hellinger distance score for the k1 node.
+
+        count_k2 = (\sqrt{\frac{N_{right_0}}{N_{parent_0}}}-\sqrt{\frac{N_{right_1}}{N_{parent_1}}})^2
+
+    be the partial hellinger distance score for the k2 node.
+
+    Hellinger distance score is then defined as:
+
+        hellinger_distance = \sqrt{count_k1+count_k2}
+    """
+    cdef double node_impurity(self) nogil:
+        """Evaluate the impurity of the current node.
+        
+        Evaluate the Hellinger distance criterion as impurity of the current node,
+        i.e. the impurity of samples[start:end]. The smaller the impurity the
+        better.
+        """
+        cdef:
+            double impurity_left = 0.0
+            double impurity_right = 0.0
+            double impurity_sum = 0.0
+
+        # get the partial hellinger score, separately per child
+        self.children_impurity(&impurity_left, &impurity_right)
+
+        # hellinger score is sqrt of children impurity sum
+        impurity_sum = impurity_right + impurity_left
+        if impurity_sum >= 0:
+            impurity_sum = sqrt(impurity_sum)
+
+        return impurity_sum
+
+    cdef void children_impurity(self, double* impurity_left,
+                                double* impurity_right) nogil:
+        """Evaluate the impurity in children nodes.
+        
+        i.e. the impurity of the left child (samples[start:pos]) and the
+        impurity the right child (samples[pos:end]) using Hellinger distance.
+
+        Parameters
+        ----------
+        impurity_left : double pointer
+            The memory address to save the impurity of the left node to
+        impurity_right : double pointer
+            The memory address to save the impurity of the right node to
+        """
+        cdef:
+            # taking only first sum_left and sum_right because only single label is supported
+            double[::1] sum_left = self.sum_left[0]
+            double[::1] sum_right = self.sum_right[0]
+
+            # sum of same class in both children nodes
+            double sum_k1 = sum_left[0] + sum_right[0]
+            double sum_k2 = sum_left[1] + sum_right[1]
+
+            # hellinger distance score calculation helpers
+            double count_k1_left = 0.0
+            double count_k2_left = 0.0
+            double count_k1_right = 0.0
+            double count_k2_right = 0.0
```

Comment:
I think those better be named `weight_k{1,2}_{left,right}` than `count_k{1,2}_{left,right}` as they are real numbers whose support is $[0, 1]$.

Reply:
agree
changed